src/mem/impls.rs

```diff
@@ -1,27 +1,221 @@
+use core::intrinsics::likely;
+
+const WORD_SIZE: usize = core::mem::size_of::<usize>();
+const WORD_MASK: usize = WORD_SIZE - 1;
+
+const WORD_COPY_THRESHOLD: usize = if 2 * WORD_SIZE > 16 {
+    2 * WORD_SIZE
+} else {
+    16
+};
+
+#[inline(always)]
+unsafe fn copy_forward_bytes(mut dest: *mut u8, mut src: *const u8, n: usize) {
+    let dest_end = dest.add(n);
+    while dest < dest_end {
+        *dest = *src;
+        dest = dest.add(1);
+        src = src.add(1);
+    }
+}
+
 #[inline(always)]
-pub unsafe fn copy_forward(dest: *mut u8, src: *const u8, n: usize) {
-    let mut i = 0;
-    while i < n {
-        *dest.add(i) = *src.add(i);
-        i += 1;
+unsafe fn copy_forward_aligned_words(dest: *mut u8, src: *const u8, n: usize) {
```

Comment:
It could, but I uses `u8` pointer to keep the signature identical to other functions.

I thought about using `u64` or `u128` here but it turns out to be bad for a few reason:
* Many platforms, especially those which don't have vectorization, can do at most usize load per instruction.
* It requires more alignment than usize, so more bytes have to be copied with bytewise copy.
* It makes the misalignment scenario bad. Shifting and reassemble 2*usize actually generates worse code than 1*usize.

However, one optimisation that I haven't include in this PR is to unroll the loop for the aligned scenario. If n is a multiple of 8*usize for example, then we can have a 8-unrolled loop followed by a loop for the remainder. On clang this could be done easily with `#pragma unroll` but I don't think it is currently possible in Rust so we might have to be it manually.

Reply:
I feel like taking `usize` here would make the intent clearer where it's copying based on each word. Perhaps there could be a typedef in the module for the copy size? Some platforms may have better performance on some types than others, so I could see it being conditional as well.

For unrolling I would expect LLVM to do this automatically if supported.